src/plugin_discovery/plugin_spec/plugin_spec.js

```diff
@@ -0,0 +1,176 @@
+import { resolve, basename, isAbsolute as isAbsolutePath } from 'path';
+
+import toPath from 'lodash/internal/toPath';
+import { get } from 'lodash';
+
+import { createInvalidPluginError } from '../errors';
+import { isVersionCompatible } from './is_version_compatible';
+
+export class PluginSpec {
+  /**
+   *  @param {PluginPack} pack - The plugin pack that produced this spec
+   *  @param {Object} opts - the options for this plugin
+   *  @param {String} [opts.id=pkg.name] - the id for this plugin.
+   *  @param {Object} [opts.uiExports] - a mapping of UiExport types
+   *                                   to UI modules or metadata about
+   *                                   the UI module
+   *  @param {Array} [opts.require] - the other plugins that this plugin
+   *                                requires. These plugins must exist and
+   *                                be enabled for this plugin to function.
+   *                                The require'd plugins will also be
+   *                                initialized first, in order to make sure
+   *                                that dependencies provided by these plugins
+   *                                are available
+   *  @param {String} [opts.version=pkg.version] - the version of this plugin
+   *  @param {Function} [opts.init] - A function that will be called to initialize
+   *                                this plugin at the appropriate time.
+   *  @param {Function} [opts.configPrefix=this.id] - The prefix to use for configuration
+   *                                values in the main configuration service
+   *  @param {Function} [opts.config] - A function that produces a configuration
+   *                                  schema using Joi, which is passed as its
+   *                                  first argument.
+   *  @param {String|False} [opts.publicDir=path + '/public']
+   *     - the public directory for this plugin. The final directory must
+   *     have the name "public", though it can be located somewhere besides
+   *     the root of the plugin. Set this to false to disable exposure of a
+   *     public directory
```

Comment:
nitpick: I think it reads better when we consistently break on 80 lines, plus use the entire newlines for the descriptions (And also remove the unnecessary `-` from the descriptions).

```js
/**
 * @param {PluginPack} pack The plugin pack that produced this spec
 * @param {Object} opts The options for this plugin
 * @param {String} [opts.id=pkg.name] The id for this plugin.
 * @param {Object} [opts.uiExports] A mapping of UiExport types to UI modules
 * or metadata about the UI module
 * @param {Array} [opts.require] The other plugins that this plugin requires.
 * These plugins must exist and be enabled for this plugin to function. The
 * require'd plugins will also be initialized first, in order to make sure
 * that dependencies provided by these plugins are available.
 * @param {String} [opts.version=pkg.version] The version of this plugin
 * @param {Function} [opts.init] A function that will be called to initialize
 * this plugin at the appropriate time.
 * @param {Function} [opts.configPrefix=this.id] The prefix to use for
 * configuration values in the main configuration service
 * @param {Function} [opts.config] A function that produces a configuration
 * schema using Joi, which is passed as its first argument.
 * @param {String|False} [opts.publicDir=path + '/public'] The public directory
 * for this plugin. The final directory must have the name "public", though it
 * can be located somewhere besides the root of the plugin. Set this to false
 * to disable exposure of a public directory.
 */
```

This is what it looks like highlighted:

<img width="768" alt="screen shot 2017-11-13 at 13 14 03" src="https://user-images.githubusercontent.com/5259/32725252-a3d4b61e-c874-11e7-95fe-bb71ccb7f5c3.png">